File: 2434-UsingaRobottoPrinttheLexicographicallySmallestString/2434-UsingaRobottoPrinttheLexicographicallySmallestString.py
class Solution:
    def robotWithString(self, s: str) -> str:
        #Initialize the length of the given word the robot array and the paper answer
        n, suffix, p = len(s), [None] * len(s), ""

        # created a suffix min
        for i in range(n - 1, -1, -1):
            if i == n - 1 or s[i] < suffix[i + 1]:
                suffix[i] = s[i]
            else:
                suffix[i] = suffix[i + 1]
        
        t = []
        
        # operations from s to t and from t to p
        for i in range(n):
            t.append(s[i])

            while t and (i == n - 1 or t[-1] <= suffix[i + 1]):
                p += t.pop()
        
        return p

        

File: 2434-UsingaRobottoPrinttheLexicographicallySmallestString/2434-UsingaRobottoPrinttheLexicographicallySmallestString.js
/**
 * @param {string} s
 * @return {string}
 */
var robotWithString = function(s) {
    //declare the robot and paper stacks
    let robotStack = [];
    let paperStack = [];
    let n = s.length;
    let suffixMin = new Array(n).fill(null);

    suffixMin[n - 1] = s[n - 1]

    for( let i = n - 2; i >= 0; i--){
        if ( s[i] < suffixMin[i + 1]){
            suffixMin[i] = s[i];
        } else{
            suffixMin[i] = suffixMin[i + 1];
        }
    }

    for(let i = 0; i < n; i++){
        robotStack.push(s[i]);

        while(robotStack.length > 0 && (i === n -1 || robotStack[robotStack.length - 1] <= suffixMin[i + 1] )){
            paperStack.push(robotStack.pop());
        }
    }


    //return the paper stack  
    return paperStack.join('')
};